### PR TITLE
Make LinearAdapter a nn.Linear child to maintain ckpt structure

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/peft/lora.py [hunk 1]
```diff
@@ -54,46 +66,117 @@ def __init__(
...
+            bias=orig_linear.bias is not None,
+            device=orig_linear.weight.device,
+            dtype=orig_linear.weight.dtype,
+        )
+        LinearAdapter._init_adapter(self, orig_linear)
 
-        self.orig_linear = orig_linear
-        self.dim = dim
-        self.scale = alpha / dim
+    @staticmethod
```
Comment (cuichenx) [inline nemo/collections/llm/peft/lora.py:79]: i think this should be a classmethod to use cls as first variable?

File: nemo/collections/llm/peft/lora.py [hunk 2]
```diff
@@ -54,46 +66,117 @@
...
+            x = obj.dropout(x)
+        lora_res = x @ obj.lora_a
+        lora_res = lora_res @ obj.lora_b
+        lora_res = lora_res * obj.scale
+        if obj.dropout_position == 'post':
+            lora_res = obj.dropout(lora_res)
         return res + lora_res
 
+    def forward(self, x):
+        return self._forward(x)
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/collections/llm/peft/lora.py:133]: ditto, shouldn't _forward be a normal method with `self` as the first arg?

File: nemo/collections/llm/peft/lora.py [hunk 3]
```diff
@@ -168,13 +251,19 @@ def transform(self, m: nn.Module, name=None, prefix=None):
...
-                    alpha=self.alpha,
-                    dropout=self.dropout,
-                    lora_A_init_method=self.lora_A_init_method,
-                    lora_dtype=self.lora_dtype,
+                    dim=8,
+                    alpha=32,
+                    dropout=0.1,
+                    dropout_position='post',
+                    lora_A_init_method='xavier',
+                    lora_dtype=None,
```
Comment (cuichenx) [inline nemo/collections/llm/peft/lora.py:266]: were these temporary?

Comment (akoumpa) [reply nemo/collections/llm/peft/lora.py]: fixed, thanks.

File: nemo/collections/llm/peft/lora.py [hunk 4]
```diff
@@ -54,46 +66,117 @@ def __init__(
...
+    def _init_adapter(
+        obj,
+        orig_linear=None,
+        dim=8,
+        alpha=32,
+        dropout=0.1,
+        dropout_position='post',
+        lora_A_init_method='xavier',
+        lora_dtype=None,
+    ):
```
Comment (cuichenx) [inline nemo/collections/llm/peft/lora.py:89]: I think it would be great if you can explain when is this function called with orig_linear = None, and when it's called with orig_linear not None. That will motivate the use of staticmethod with this `obj` argument

File: nemo/collections/llm/peft/lora.py [hunk 5]
```diff
@@ -54,46 +66,117 @@ def __init__(
...
-        dtype = lora_dtype or self.orig_linear.weight.dtype
-
-        self.lora_a = nn.Parameter(torch.zeros((in_features, dim), dtype=dtype, device=device))
-        self.lora_b = nn.Parameter(torch.zeros((dim, out_features), dtype=dtype, device=device))
+        device = obj.weight.device
+        obj.weight.requires_grad = False
+        if obj.bias is not None:
+            obj.bias.requires_grad = False
+        # copy weights
+        if orig_linear is not None:
```
Comment (akoumpa) [inline nemo/collections/llm/peft/lora.py:99]: In the FSDPv1 path, orig_linear has not been allocated yes (operates within a `device("meta")` context), therefore we cannot copy any weights. Instead this function monkey-patches the input obj (in this case an `nn.Linear`)

**Review by cuichenx — APPROVED** [review]
Approving code changes with a request on comments

**Review by cuichenx — APPROVED** [review]
LGTM

File: nemo/collections/llm/peft/lora.py (hunk 1 above)
Comment (cuichenx) [reply nemo/collections/llm/peft/lora.py:87]: resolved offline

File: .github/workflows/cicd-main.yml [hunk 6]
```diff
@@ -3611,6 +3611,17 @@ jobs:
...
+  L2_VLM_HF_Transformer_PEFT_FSDP:
+    needs: [ cicd-test-container-setup ]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_VLM_HF_Transformer_PEFT_FSDP') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+        TRANSFORMERS_OFFLINE=1 python tests/collections/vlm/hf/peft.py --model /home/TestData/vlm/qwen2-2b/ --max-steps 3 --disable-ckpt --strategy fsdp --devices 2
+      AFTER_SCRIPT: |
+        rm -rf nemo_experiments
```
Comment (pablo-garay) [inline .github/workflows/cicd-main.yml:3623]: i dont think this line is needed?

**Review by pablo-garay — APPROVED** [review]
